test/integration/extensions_test.go

```diff
@@ -0,0 +1,229 @@
+package integration
+
+import (
+	"bytes"
+	"crypto/tls"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	configapi "github.com/openshift/origin/pkg/cmd/server/api"
+	testutil "github.com/openshift/origin/test/util"
+)
+
+func TestExtensions(t *testing.T) {
+	// Create a temporary directory.
+	tmpdir, err := ioutil.TempDir("", "extensions")
+	if err != nil {
+		t.Fatalf("Could not create tmp dir for extensions: %v", err)
+	}
+	defer os.RemoveAll(tmpdir)
+
+	// Create extension files.
+	masterOptions := setupExtensions(t, tmpdir)
+
+	// Start server.
+	_, err = testutil.StartConfiguredMaster(masterOptions)
+	if err != nil {
+		t.Fatalf("Unexpected error starting server: %v", err)
+	}
+
+	// Request the extension scripts.
+	expectedScripts := concat(";\n", script1, script2)
+	testURL := masterOptions.AssetConfig.PublicURL + "scripts/extensions.js"
+	testGET(t, testURL, http.StatusOK, "text/javascript", expectedScripts)
+
+	// Request the extension stylesheets.
+	expectedStylesheets := concat("\n", stylesheet1, stylesheet2)
+	testURL = masterOptions.AssetConfig.PublicURL + "styles/extensions.css"
+	testGET(t, testURL, http.StatusOK, "text/css", expectedStylesheets)
+
+	// Inject the base into index.html to test HTML5Mode
+	publicURL, err := url.Parse(masterOptions.AssetConfig.PublicURL)
+	if err != nil {
+		t.Fatalf("Unexpected error parsing PublicURL %q: %v", masterOptions.AssetConfig.PublicURL, err)
+	}
+	baseInjected := injectBase(indexHTML, "extension1", publicURL)
+
+	// With HTML5Mode on, extension1 should have the base URI injected into index.html.
+	testURL = masterOptions.AssetConfig.PublicURL + "extensions/extension1/"
+	testGET(t, testURL, http.StatusOK, "text/html", baseInjected)
+
+	testURL = masterOptions.AssetConfig.PublicURL + "extensions/extension1/files/shakespeare.txt"
+	testGET(t, testURL, http.StatusOK, "text/plain", []byte(plaintext))
+
+	testURL = masterOptions.AssetConfig.PublicURL + "extensions/extension1/does-not-exist/"
+	testGET(t, testURL, http.StatusOK, "text/html", baseInjected)
+
+	// With HTML5Mode off, extension2 should NOT have the base URI injected.
+	testURL = masterOptions.AssetConfig.PublicURL + "extensions/extension2/"
+	testGET(t, testURL, http.StatusOK, "text/html", []byte(indexHTML))
+
+	testURL = masterOptions.AssetConfig.PublicURL + "extensions/extension2/does-not-exist/"
+	testGET(t, testURL, http.StatusNotFound, "", nil)
+
+	// 404 since the file is in extension1/files
+	testURL = masterOptions.AssetConfig.PublicURL + "extensions/extension2/files/shakespeare.txt"
+	testGET(t, testURL, http.StatusNotFound, "", nil)
+}
+
+func setupExtensions(t *testing.T, tmpdir string) *configapi.MasterConfig {
+	check := func(err error) {
```

Comment:
this is more concise, but makes it hard to see which line an unexpected error originated on, since the stack will show line 79, not the real source